File: cpp/examples/test3.cpp
#include <wup/wup.hpp>

using namespace wup;

int
main() {
	srand(time(NULL));

	Dataset dataset("../../datasets/uji2");
    
    node::Root         root(dataset.numFeatures());
    node::ZScore       zscore1(&root, true);
    node::Smooth       smooth(&zscore1, 0.01);
    node::Direction    direction(&smooth);
    node::ZScore       zscore2(&direction, seq<int>(0,1), false);
    node::Rotate       rotate(&zscore2);
    node::Tanh         tanh(&rotate, seq<int>(0,1));
    node::Replicate    replicate(&tanh, 3);
    node::KernelCanvas kernelCanvas(&replicate, 1024, 0.01, 16);

    root.addEmitter(&kernelCanvas);

    Wisard w(root.binaryOutputLength(), 32, dataset.classes());

    const int length = dataset.size();
    const int threshold = length * 0.9;
    const int * const indexes = randperm(length);

    print("Training...");
//    for (int degrees=-5;degrees<=+5;degrees+=5) {
//    	print("Degrees:", degrees);
//    	rotate.angle(degrees);

    	Clock trainTime;
		for (int i=0;i<threshold;++i) {
			if (i%10==0) printn("\r",i,"/",threshold);
			Sample & sample = dataset[indexes[i]];
			w.learn(root.digest(sample), sample.target());
		}
		trainTime.stop();
		print(", UnitTrainTime(ms): ", trainTime.ellapsed() / double(threshold) / 1000.0);
//    }

    print();

    for (int c=0;c<dataset.classes();++c) {
		print("Testing...");
		int hits=0, misses=0;
		Clock testTime;
		for (int i=threshold;i<length;++i) {
			if (i%10==0) printn("\r",i-threshold,"/",length-threshold);
			Sample & sample = dataset[indexes[i]];
			if (w.readBleaching(root.digest(sample)) == sample.target())
				++hits;
			else
				++misses;
		}
		testTime.stop();
		print(", UnitTestTime(ms): ", testTime.ellapsed() / double(length-threshold) / 1000.0);

		print("Accuracy: ", hits / double(hits + misses));

		Clock forgetTime;
		w.forget(c);
		forgetTime.stop();
		print("TimeToForget(ms): ", forgetTime.ellapsed() / 1000.0);
    }

    return 0;
}

File: android/AndWUPExamples/src/com/github/diegofps/android/andwup/examples/Test3.java
package com.github.diegofps.android.andwup.examples;

import com.github.diegofps.android.andwup.models.Wisard;
import com.github.diegofps.android.andwup.nodes.Direction;
import com.github.diegofps.android.andwup.nodes.KernelCanvas;
import com.github.diegofps.android.andwup.nodes.Replicate;
import com.github.diegofps.android.andwup.nodes.Root;
import com.github.diegofps.android.andwup.nodes.Rotate;
import com.github.diegofps.android.andwup.nodes.Smooth;
import com.github.diegofps.android.andwup.nodes.Tanh;
import com.github.diegofps.android.andwup.nodes.ZScore;

public class Test3 extends Thread {

	@Override
	public void run() {
		int[] columns = new int[] {0, 1};
		
	    Root         root         = new Root(2);
	    ZScore       zscore1      = new ZScore(root, true);
	    Smooth       smooth       = new Smooth(zscore1, 0.01);
	    Direction    direction    = new Direction(smooth);
	    ZScore       zscore2      = new ZScore(direction, columns, false);
	    Rotate       rotate       = new Rotate(zscore2);
	    Tanh         tanh         = new Tanh(rotate, columns);
	    Replicate    replicate    = new Replicate(tanh, 3);
	    KernelCanvas kernelCanvas = new KernelCanvas(replicate, 1024, 0.01, 16);

	    root.addEmitter(kernelCanvas);

	    Wisard w = new Wisard(root.binaryOutputLength(), 32, 2);

	    
	    
	}
	
}
